### PR TITLE
Vecsim norm opt

### PR DESCRIPTION
**Pull request by GuyAv46**
followup for [this PR](https://github.com/RedisAI/VectorSimilarity/pull/145) on VectorSimilarity, now we normalize the query vector only once and if needed, then using `getDistanceFrom` that now does not normalize by itself

### CODE REVIEW

File: src/hybrid_reader.h [hunk 1]
```diff
@@ -18,11 +18,14 @@ typedef struct {
   IndexIterator base;
   VecSimIndex *index;
   KNNVectorQuery query;
-  VecSimQueryParams runtimeParams;   // Evaluated runtime params.
+  size_t dimention;                // index dimention
```
Comment (DvirDukhan) [inline src/hybrid_reader.h:21]: typo
```suggestion
  size_t dimension;                // index dimension
```

File: src/hybrid_reader.h [hunk 2]
```diff
@@ -37,7 +40,7 @@ typedef struct {
 extern "C" {
 #endif
 
-IndexIterator *NewHybridVectorIterator(VecSimIndex *index, char *score_field, KNNVectorQuery query, VecSimQueryParams qParams, IndexIterator *child_it);
+IndexIterator *NewHybridVectorIterator(VecSimIndex *index, char *score_field, KNNVectorQuery query, size_t dim, VecSimType type, VecSimMetric metric, VecSimQueryParams qParams, IndexIterator *child_it);
```
Comment (DvirDukhan) [inline src/hybrid_reader.h:43]: can you pass the &info instead of a big list of parameters?

File: tests/pytests/test_vecsim.py [hunk 3]
```diff
@@ -792,3 +792,53 @@ def test_wrong_vector_size(env):
...
+           'RETURN', 1, '__v_score')
+    prefix = "_" if env.isCluster() else ""
+    env.assertEqual(env.cmd(prefix+"FT.DEBUG", "VECSIM_INFO", "idx", "v")[-1], 'HYBRID_BATCHES')
+    # The order of ids is not accurate due to floating point numeric errors, but the top k should be
+    # the last 10 ids, and their scores should all be in the right COSINE range.
+    actual_res_ids = set([res[1:][2*i] for i in range(10)])
+    actual_res_scores = [res[1:][2*i + 1][1] for i in range(10)]
+    env.assertEqual(actual_res_ids, expected_res_ids)
+    for score in actual_res_scores:
+        env.assertTrue(0.0 <= float(score) <= 2.0)
```
Comment (DvirDukhan) [inline tests/pytests/test_vecsim.py:825]: you can drop this. it is meaningless. if we want to verify the correctness of the scores (e.g. between [0,2]) we should do it in the library

**Review by DvirDukhan — COMMENTED** [review]
👍 
look at the comments

File: src/hybrid_reader.h (hunk 2 above)
Comment (GuyAv46) [reply src/hybrid_reader.h:43]: I can, but then we will need a switch inside `NewHybridVectorIterator` again and we only need this three values

File: src/hybrid_reader.c [hunk 4]
```diff
@@ -127,9 +127,16 @@ void computeDistances(HybridIterator *hr) {
   RSIndexResult *cur_res = hr->base.current;
   RSIndexResult *cur_child_res;  // This will use the memory of hr->child->current.
   RSIndexResult *cur_vec_res = NewDistanceResult();
+  void *qvector = hr->query.vector;
+
+  if (hr->indexMetric == VecSimMetric_Cosine) {
+    qvector = rm_malloc(hr->dimension * VecSimType_sizeof(hr->vecType));
+    memcpy(qvector, hr->query.vector, hr->dimension * VecSimType_sizeof(hr->vecType));
```
Comment (alonre24) [inline src/hybrid_reader.c:134]: Why do you need to copy? can't you just normalise `hr->query.vector` in place?

**Review by alonre24 — COMMENTED** [review]
Nice, one comment

File: src/hybrid_reader.c (hunk 4 above)
Comment (GuyAv46) [reply src/hybrid_reader.c:134]: We previously decided that the blob in the PARAMS part is a read only. Assume that when we support multiple VSS in one query, client can pass the same blob for L2 and Cosine indexes

**Review by DvirDukhan — APPROVED** [review]

**Review by DvirDukhan — APPROVED** [review]

**Review by alonre24 — COMMENTED** [review]
LGTM. Look if we need to update the `HybridIteratorParam` in `benchmark_vecsim_hybrid_queries.cpp` as well

**Comment by GuyAv46** [comment]
> LGTM. Look if we need to update the `HybridIteratorParam` in `benchmark_vecsim_hybrid_queries.cpp` as well

Updated this as well

**Review by alonre24 — APPROVED** [review]

**Review by ashtul — APPROVED** [review]

**Pull request merged**